### PR TITLE
enhanced gas station registry consistency & maintenance features

### LINKED ISSUE #144: Improve Gas Splitting Algorithm Robustness and Documentation for `target_init_balance` Changes
**Description:**
The current gas-splitting algorithm in the gas station uses three thresholds to determine when and how gas coins are split:

* **T** = `target_init_balance` — the target gas coin size.
* **M** = `target_init_balance * 200` — the threshold above which a gas coin is considered *large enough* to be split.
* **S** = `2 * (avg_split_fee + target_init_balance)` — the threshold below which a gas coin is considered *too small* to split.

The algorithm is self-organizing and relies on the assumption that **T < M < S**. However, when `target_init_balance` is set too low, it may cause **M' > S'**,  or **S' > M' << M** leading to unpredictable behavior such as:

* Continuous re-adding of the same coins into the pool.
* Coins not being split properly according to the new `target_init_balance`.
* Imbalanced or inconsistent gas object sizes.

Changing `target_init_balance` can therefore introduce inconsistencies, double-added gas objects, or unstable pool behavior.

**Proposed Solution:**

1. **Short term:**

   * Update documentation to explicitly state that any change to `target_init_balance` **must be accompanied by a Redis cleanup** to prevent inconsistencies.

2. **Long term:**

   * Investigate improvements or redesign of the gas-splitting algorithm to make it more resilient to threshold imbalance or misconfiguration.
   * Consider mechanisms that dynamically adapt to configuration changes without manual cleanup.

### PR DESCRIPTION
**Pull request by qrayven**
### Description

This PR introduces several important features to improve the reliability, maintainability, and operational flexibility of the IOTA Gas Station. The changes focus on preventing registry inconsistencies, enabling safe configuration changes, and providing better multi-instance support.

#### 1. Cold Parameters System

The gas station now tracks "cold parameters" — configuration values that require a full coin registry rescan when changed. Currently, this includes the `target_init_balance` setting.

When the gas station starts, it compares the current configuration against previously stored cold parameters in Redis. If any changes are detected, the system requires explicit acknowledgment via the `--allow-reinit` flag before proceeding with a full rescan. This prevents accidental misconfiguration from silently corrupting the coin registry.

#### 2. Storage Namespaces

Redis keys are now prefixed with a network-aware namespace that combines the fullnode host, port, and sponsor address. This prevents data collisions when running multiple gas station instances against different networks using the same Redis instance. For example, a devnet instance and a testnet instance can now safely share the same Redis database without interfering with each other.

#### 3. Automatic Key Migration

To ensure backward compatibility, the system automatically detects and migrates data from the old key format to the new namespaced format on startup. The migration runs atomically via Lua scripts, preserving any existing data in the new namespace. The process logs detailed statistics about how many keys were migrated, skipped, or encountered errors.

#### 4. Maintenance Mode

A distributed maintenance lock mechanism now protects registry operations across multiple gas station instances. When performing a full rescan or registry cleanup, the system first acquires an exclusive maintenance lock in Redis. While this lock is held, all coin reservation requests from any instance return a clear error message indicating the gas station is temporarily unavailable. The lock includes a time-to-live to prevent permanent deadlocks if an instance crashes during maintenance.

#### 5. Consistency Checks

Before each scheduled rescan cycle, the gas station now performs a quick consistency check comparing the Redis registry state against the actual network state. It checks both the total coin count and aggregate balance, using configurable divergence thresholds (defaulting to 20%). If the registry has drifted significantly from reality — perhaps due to external transfers or registry corruption — the system logs detailed warnings with actionable remediation steps, suggesting the use of `--allow-reinit` and `--delete-coin-registry` flags.

#### 6. Target Init Balance Validation

A new safety check validates that the configured `target_init_balance` is large enough relative to network gas costs. This prevents creating coins so small that gas fees would consume them entirely, which would lead to registry inconsistencies. If the validation fails, initialization stops with a clear error message explaining the minimum required balance.

#### 7. Delete Registry CLI Flag

A new `--delete-coin-registry` flag allows operators to force a complete clean re-initialization of the coin registry. When combined with `--allow-reinit`, this triggers maintenance mode, deletes all existing registry data, and performs a fresh scan of on-chain coins. This is useful for recovering from corrupted registry states or when making significant configuration changes.

### CLI Changes

Three new flags have been added:
- `--allow-reinit` (`-a`): Permits automatic reinitialization when cold parameters change
- `--delete-coin-registry` (`-d`): Forces deletion of the existing registry before starting (requires `--allow-reinit`)
- `--ignore-locks`: Ignores the maintenance lock and initialization lock. 

### Breaking Changes

- **Redis key format changed**: Keys now include a network namespace prefix. The automatic migration handles existing data transparently, but operators should be aware of this structural change.
- **Maintenance mode behavior**: During full rescans, coin reservation requests will temporarily fail with a descriptive error rather than silently waiting or returning stale data.

### Testing

Added unit tests for consistency check calculations including edge cases with zero values and large numbers. Added integration tests covering various Redis migration scenarios including partial migrations and key conflicts. Added tests for the new aggregate coin stats fetching functionality used by consistency checks.

closes #144

### CODE REVIEW

File: src/config/cold_params.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            ));
+        }
+        changes
+    }
+}
+
+pub type ColdParamChange = String;
+
+/// Checks if the cold params have changes compared to the storage and returns the changes.
+pub async fn check_cold_params_changes(
```
Comment (wulfraem) [inline src/config/cold_params.rs]: Would it make sense to add this function to the `ColdParams` struct? Kinda looking like it belongs there, as the first argument is an instance of it, and could also be just name `self`. :D

File: src/config/cold_params.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    let maybe_cold_params = storage
+        .get_data(key)
+        .await
+        .context("unable to get cold params from storage")?;
+
+    // If there are no cold params, it means this is the first run.
+    // In this case, there are no changes, so we only need to save the current cold params.
+    if maybe_cold_params.is_none() {
+        storage
+            .set_data(key, serde_json::to_vec(current_cold_params).unwrap())
```
Comment (wulfraem) [inline src/config/cold_params.rs]: Are the 3 unwraps here leftovers from an initial version of this function? ^^

File: src/gas_station_initializer.rs [hunk 3]
```diff
@@ -304,14 +369,30 @@ impl GasStationInitializer {
...
             .calibrate_gas_cost_per_object(sponsor_address, &coins[0])
             .await;
         info!("Calibrated gas cost per object: {:?}", gas_cost_per_object);
+
+        // Safety check: Validate minimum coin size prevent registry inconsistency
+        if target_init_coin_balance * 99 <= gas_cost_per_object {
+            let err_msg = format!(
+                "target_init_coin_balance ({}) is too small relative to gas_cost_per_object ({}). \
+                 Minimum required: target_init_coin_balance > gas_cost_per_object / 99 = {}. \
+                 Please increase target_init_coin_balance to prevent registry inconsistency.",
```
Comment (wulfraem) [inline src/gas_station_initializer.rs]: Could we change the message slightly to avoid confusion? At least it confused me a bit. ^^;;

The line where a minimal value is provided, kinda confused me, as it looks like a formula that doesn't make sense (`A > B = C`).

Maybe something like this?


```suggestion
                "target_init_coin_balance ({}) is too small relative to gas_cost_per_object ({}). \
                 target_init_coin_balance must be greater than gas_cost_per_object / 99 \
                 to prevent registry inconsistencies. Please increase target_init_coin_balance  \
                 to at least {} for current network and coin setup.",
```

File: README.md [hunk 4]
```diff
@@ -121,8 +121,11 @@ access-controller:
...
 
-> [!WARNING]
-> **Important:** Some configuration parameters require the Redis database to be cleared (flushed) before changes can take effect safely. Modifying these settings without resetting the database may result in inconsistencies or corruption of the gas coin registry. **Always flush the Redis database prior to changing such parameters and restart the service to avoid issues.**
+
+#### Gas Station reinitialization
+
+Certain configuration parameters require the Redis database to be cleared (flushed) before any changes to those settings can take effect safely. If you modify these parameters, you will typically be notified that a reinitialization is required. To prevent accidental or unintended reinitializations—which may take a significant amount of time—you must explicitly start the gas station with the `--allow-reinit` flag to allow automatic reinitialization. Alternatively, you can revert the changed parameters to their original values and plan the reinitialization for a more convenient time.
```
Comment (wulfraem) [inline README.md]: Could we list those here, so that users know them when updating the config? ^^

File: src/command.rs [hunk 5]
```diff
@@ -31,11 +32,28 @@ use tracing::info;
...
     #[arg(env, long, help = "Path to config file")]
     config_path: PathBuf,
+    #[arg(
+        env,
+        short,
+        long,
+        help = "Allow reinitialization of the gas station when the target init balance parameter changes",
+        default_value_t = false
+    )]
+    allow_reinit: bool,
```
Comment (wulfraem) [inline src/command.rs:49]: Could a maintenance lock be set opened by an instance, that for some unknown reason crashes due to an unpredicted reason (e.g. due to something from within the gas station, or just a server restart), and then block the gas station start with `force_full_rescan` enabled for 12 hours?

Was thinking if it would make sense to add a new flag, that allows to override such values in the Redis. But the wording might become a bit more difficult, e.g. we could call this "override" instead of "force" to avoid confusion, as the entire reinit step is optional anyway, as both flags would allow such a reinit and not enforce it. Then split the current allow into a lowercase and an uppercase variant (also shifted to "r"/"R" for instead of "a" for the shorthand):

```Rust
    #[arg(
        env,
        short = 'r',
        long,
        help = "Allow reinitialization of the gas station when the target init balance parameter changes. This flag will be ignored, if another Gas Station instance is already performing a reinitialization.",
        default_value_t = false
    )]
    allow_reinit: bool,
    #[arg(
        env,
        short = 'R',
        long,
        help = "Allow reinitialization of the gas station when the target init balance parameter changes. This flag will attempt to run a reinitialization in case another attempt failed and left it blocked. Use this flag only if you are sure, no other instance is already performing a reininitialization, as this process is not intended to run on multiple connected instances the same time!",
        default_value_t = false
    )]
    override_reinit: bool,
```

File: src/storage/redis/mod.rs [hunk 6]
```diff
@@ -232,12 +300,49 @@ impl Storage for RedisStorage {
...
+        debug!("Releasing the maintenance lock.");
+        let mut conn = self.conn_manager.clone();
+        ScriptManager::release_maintenance_lock_script()
+            .arg(self.namespace.clone())
             .invoke_async::<_, ()>(&mut conn)
             .await?;
         Ok(())
     }
 
+    async fn is_maintenance_mode(&self) -> anyhow::Result<bool> {
```
Comment (wulfraem) [inline src/storage/redis/mod.rs:331]: Quite sure, I'm just overlooking something, but is this actually used outside the tests below in this file?

Related to this, so going to put the question here: Do we have a distinct return value (HTTP code and/or message) for maintenance mode? Seems like we're using the default error handling at the moment, which would return an error 500.

Shold we consider returning an [error 503](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/503) instead? IMO, it might be a better fit, but might be a bit too refactoring for this issue, and if we actually want different return codes, we could also create a refactor issue to add this in the future. From a short peak, even an invalid argument would return error 500 (from a short look at `reserve_gas_impl`), so it might already be an interesting point to improve.

**Review by wulfraem — CHANGES_REQUESTED** [review]

File: src/config/cold_params.rs (hunk 1 above)
Comment (qrayven) [reply src/config/cold_params.rs]: good point, fixed

File: src/config/cold_params.rs (hunk 2 above)
Comment (qrayven) [reply src/config/cold_params.rs]: yes, this is true ;) fixed

File: README.md (hunk 4 above)
Comment (qrayven) [reply README.md]: done

File: src/storage/redis/mod.rs (hunk 6 above)
Comment (qrayven) [reply src/storage/redis/mod.rs:331]: Good point,  fixed. The implementation isn't beatiful but I think this is acceptable?

Comment (wulfraem) [reply src/storage/redis/mod.rs:331]: That should work, and it avoids, having to refactor the entire error handling. ^^

File: README.md [hunk 7]
```diff
@@ -121,8 +121,10 @@ access-controller:
...
 
-> [!WARNING]
-> **Important:** Some configuration parameters require the Redis database to be cleared (flushed) before changes can take effect safely. Modifying these settings without resetting the database may result in inconsistencies or corruption of the gas coin registry. **Always flush the Redis database prior to changing such parameters and restart the service to avoid issues.**
+
+#### Gas Station reinitialization
+
+The configuration parameter `target-init-balance` requires the Redis database to be cleared (flushed) before any changes to those settings can take effect safely. If you modify these parameters, you will typically be notified that a reinitialization is required. To prevent accidental or unintended reinitializations—which may take a significant amount of time—you must explicitly start the gas station with the `--allow-reinit` flag to allow automatic reinitialization. Alternatively, you can revert the changed parameters to their original values and plan the reinitialization for a more convenient time.
```
Comment (wulfraem) [inline README.md]: Maybe? ^^

```suggestion
The configuration parameter `target-init-balance` requires the Redis database to be cleared (flushed) before any changes to those settings can take effect safely. If you modify these parameters, you will typically be notified that a reinitialization is required. To prevent accidental or unintended reinitializations — which may take a significant amount of time — you must explicitly start the gas station with the `--allow-reinit` flag to allow automatic reinitialization. Alternatively, you can revert the changed parameters to their original values and plan the reinitialization for a more convenient time.
```

**Review by wulfraem — CHANGES_REQUESTED** [review]

File: src/command.rs (hunk 5 above)
Comment (qrayven) [reply src/command.rs:49]: ok I've added the `--ignore-locks` and also the description in the common issues section

File: docs/common-issues.md [hunk 8]
```diff
@@ -97,3 +97,88 @@ In case you use Docker Compose, please edit the docker-compose file:
...
+
+Currently, the following parameters are cold params:
+
+- **`target_init_balance`** — The target balance for each gas coin in the pool. Changing this value requires re-splitting all coins to match the new target size.
+
+When a cold param change is detected, the Gas Station refuses to start automatically to prevent accidental data loss or extended downtime. This is a safety mechanism because:
+
+1. Full re-initialization deletes the existing coin registry and rescans all coins from the network
+2. The process can take significant time depending on the number of coins
+3. During re-initialization, the instance enters maintenance mode, which affects all other running instances
```
Comment (wulfraem) [inline docs/common-issues.md]: Should we add a note about this effect here as well?

```suggestion
3. During re-initialization, the instance enters maintenance mode, which blocks all coin reservation requests on the current and all other running instances
```

File: src/gas_station_initializer.rs [hunk 9]
```diff
@@ -225,6 +271,28 @@ impl GasStationInitializer {
         }
     }
 
+    // Performs full rescan operations while holding the maintenance lock.
```
Comment (wulfraem) [inline src/gas_station_initializer.rs]: Was the function doc reduced to an inline comment intentionally? If not, I'd be in favor of keeping it. ^^

```suggestion
    /// Performs full rescan operations while holding the maintenance lock.
```

**Review by wulfraem — CHANGES_REQUESTED** [review]

**Review by wulfraem — APPROVED** [review]
